src/Compilers/CSharp/Portable/Syntax/LambdaUtilities.cs

```diff
@@ -0,0 +1,269 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    internal static class LambdaUtilities
+    {
+        /// <summary>
+        /// Returns true if the specified node represents a lambda.
+        /// </summary>
+        public static bool IsLambda(SyntaxNode node)
+        {
+            switch (node.Kind())
+            {
+                case SyntaxKind.ParenthesizedLambdaExpression:
+                case SyntaxKind.SimpleLambdaExpression:
+                case SyntaxKind.AnonymousMethodExpression:
+                case SyntaxKind.LetClause:
+                case SyntaxKind.WhereClause:
+                case SyntaxKind.AscendingOrdering:
+                case SyntaxKind.DescendingOrdering:
+                case SyntaxKind.SelectClause:
```

Comment:
It is spelled out in detail in the C# language specification. See section 7.16.2. The translation is defined recursively. See, for example, see 7.16.2.5 for a case that generates no lambda, and note the exception given in 7.16.2.3. There are a number of special cases and the effects may be subtle due to the recursive interactions of the rules.